From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vansencool <vansencool@gmail.com>
Date: Thu, 23 Jan 2025 22:11:31 +0530
Subject: [PATCH] Add typesafe config to API


diff --git a/build.gradle.kts b/build.gradle.kts
index 336b6891d246a43f8fb66bdcda04c565cc925c7e..9331d76f6f64dcf5b5d909845b181ab1d78261c8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -89,6 +89,8 @@ dependencies {
     // commons-lang3 is removed in maven-resolver-provider since 3.9.8
     // Add this because bukkit api still need it.
     compileOnly("org.apache.commons:commons-lang3:3.17.0")
+
+    apiAndDocs("com.typesafe:config:1.4.3") // NorsPaper - Add typesafe config to API
 }
 
 // Paper start
@@ -204,7 +206,10 @@ tasks.withType<Javadoc> {
     doFirst {
         options.addStringOption(
             "sourcepath",
-            apiAndDocs.resolvedConfiguration.files.joinToString(separator = File.pathSeparator, transform = File::getPath)
+            apiAndDocs.resolvedConfiguration.files.joinToString(
+                separator = File.pathSeparator,
+                transform = File::getPath
+            )
         )
     }
 
@@ -249,12 +254,13 @@ tasks.check {
 }
 // Paper end
 // Paper start
-val scanJarForOldGeneratedCode = tasks.register("scanJarForOldGeneratedCode", io.papermc.paperweight.tasks.ScanJarForOldGeneratedCode::class) {
-    mcVersion.set(providers.gradleProperty("mcVersion"))
-    annotation.set("Lio/papermc/paper/generated/GeneratedFrom;")
-    jarToScan.set(tasks.jar.flatMap { it.archiveFile })
-    classpath.from(configurations.compileClasspath)
-}
+val scanJarForOldGeneratedCode =
+    tasks.register("scanJarForOldGeneratedCode", io.papermc.paperweight.tasks.ScanJarForOldGeneratedCode::class) {
+        mcVersion.set(providers.gradleProperty("mcVersion"))
+        annotation.set("Lio/papermc/paper/generated/GeneratedFrom;")
+        jarToScan.set(tasks.jar.flatMap { it.archiveFile })
+        classpath.from(configurations.compileClasspath)
+    }
 tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
