From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vansencool <vansencool@gmail.com>
Date: Thu, 23 Jan 2025 00:02:34 +0530
Subject: [PATCH] Changes in version


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index 13d6cd2e4062db77cc17c7c1f9facb4777c0a0db..c24584dd7c03ddd1681d8bb42fb3d5ab8ec97eaf 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -36,7 +36,7 @@ public record ServerBuildInfoImpl(
     private static final String BRAND_PUFFERFISH_NAME = "Pufferfish"; // Leaf
     private static final String BRAND_PURPUR_NAME = "Purpur"; // Purpur
     private static final String BRAND_LEAF_NAME = "Leaf"; // Leaf
-    private static final String BRAND_NORSPAPER = "NorsPaper";
+    private static final String BRAND_NORSPAPER = "NorsPaper"; // NorsPaper
 
     private static final String BUILD_DEV = "DEV";
 
@@ -48,9 +48,9 @@ public record ServerBuildInfoImpl(
         this(
                 getManifestAttribute(manifest, ATTRIBUTE_BRAND_ID)
                         .map(Key::key)
-                        .orElse(BRAND_NORSPAPER_ID), // Gale - branding changes // Leaf
+                        .orElse(BRAND_NORSPAPER_ID), // Gale - branding changes // Leaf // NorsPaper
                 getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
-                        .orElse(BRAND_NORSPAPER), // Gale - branding changes // Leaf
+                        .orElse(BRAND_NORSPAPER), // Gale - branding changes // Leaf // NorsPaper
                 SharedConstants.getCurrentVersion().getId(),
                 SharedConstants.getCurrentVersion().getName(),
                 getManifestAttribute(manifest, ATTRIBUTE_BUILD_NUMBER)
diff --git a/src/main/java/org/galemc/gale/version/AbstractPaperVersionFetcher.java b/src/main/java/org/galemc/gale/version/AbstractPaperVersionFetcher.java
index d0a031014ec410142d59c8edd577bf035b7e407b..2d171eae5d16f50d77c78d8aa8cd2273eb3e7d63 100644
--- a/src/main/java/org/galemc/gale/version/AbstractPaperVersionFetcher.java
+++ b/src/main/java/org/galemc/gale/version/AbstractPaperVersionFetcher.java
@@ -70,7 +70,7 @@ public abstract class AbstractPaperVersionFetcher implements VersionFetcher {
         final Component updateMessage;
         final ServerBuildInfo build = ServerBuildInfo.buildInfo();
         if (build.buildNumber().isEmpty() && build.gitCommit().isEmpty()) {
-            updateMessage = text("You are running a development version without access to version information", color(0xFF5300));
+            updateMessage = text("You are running a development version without access to version information (Likely IDE).", color(0xFF5300));
         } else {
             updateMessage = getUpdateStatusMessage(this.gitHubOrganizationName + "/" + this.gitHubRepoName, build); // Gale - branding changes - version fetcher
         }
